File: GithubJavaPop/app/src/main/java/br/com/githubjavapop/activity/PullRequestActivity.java
package br.com.githubjavapop.activity;

import android.content.Intent;
import android.net.Uri;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ImageView;
import android.widget.Toast;

import java.util.List;

import br.com.githubjavapop.R;
import br.com.githubjavapop.adapter.PullRequestAdapter;
import br.com.githubjavapop.entidade.PullRequest;
import br.com.githubjavapop.service.PullRequestService;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class PullRequestActivity extends AppCompatActivity {

    private static String URL;
    private static String REPOSITORIO;
    private static String USERNAME;

    private List<PullRequest> listaPullRequest;
    private PullRequestAdapter pullRequestAdapter;
    private RecyclerView recyclerPullRequest;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_pull_request);

        ImageView imagemVoltar;

        imagemVoltar = (ImageView) findViewById(R.id.imagemVoltar);

        imagemVoltar.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(PullRequestActivity.this, MainActivity.class));
            }
        });

        URL = getIntent().getStringExtra("url");
        REPOSITORIO = getIntent().getStringExtra("repositorio");
        USERNAME = getIntent().getStringExtra("username");

        recyclerPullRequest = (RecyclerView) findViewById(R.id.rccPull);
        recyclerPullRequest.setLayoutManager(new LinearLayoutManager(getApplicationContext()));
        obterPulls();
    }

    private void obterPulls() {

        Retrofit retrofit = new Retrofit.Builder().baseUrl(URL)
                .addConverterFactory(GsonConverterFactory.create()).build();
        PullRequestService pullRequestService = retrofit.create(PullRequestService.class);
        Call<List<PullRequest>> pulls = pullRequestService.getPullRequest(USERNAME, REPOSITORIO);

        pulls.enqueue(new Callback<List<PullRequest>>() {
            @Override
            public void onResponse(Call<List<PullRequest>> call, Response<List<PullRequest>> response) {
                if(response.isSuccessful()) listaPullRequest = response.body();

                pullRequestAdapter = new PullRequestAdapter(listaPullRequest);

                pullRequestAdapter.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        Intent browseIt = new Intent(Intent.ACTION_VIEW, Uri.parse(listaPullRequest
                                .get(recyclerPullRequest.getChildAdapterPosition(v)).getHtmlURL()));
                        startActivity(browseIt);
                    }
                });

                recyclerPullRequest.setAdapter(pullRequestAdapter);
            }

            @Override
            public void onFailure(Call<List<PullRequest>> call, Throwable t) {
                Toast.makeText(PullRequestActivity.this, t.getMessage(),Toast.LENGTH_LONG).show();
            }
        });
    }

}

File: app/src/main/java/br/com/rmso/popularrepositories/ui/activities/PullRequestActivity.kt
package br.com.rmso.popularrepositories.ui.activities

import android.content.Intent
import android.graphics.drawable.ColorDrawable
import android.net.Uri
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import br.com.rmso.popularrepositories.ListOnClickListener
import br.com.rmso.popularrepositories.R
import br.com.rmso.popularrepositories.model.PullRequest
import br.com.rmso.popularrepositories.retrofit.RetrofitAPI
import br.com.rmso.popularrepositories.ui.adapters.PullResquestAdapter
import br.com.rmso.popularrepositories.utils.Constants
import kotlinx.android.synthetic.main.activity_pull_request.*
import kotlinx.android.synthetic.main.activity_repository.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class PullRequestActivity : AppCompatActivity(), ListOnClickListener {
    private var retrofit = RetrofitAPI()
    private var owner = ""
    private var repositoryName = ""
    private var pullRequestArrayList = ArrayList<PullRequest>()
    private val constans = Constants()
    var linearLayoutManager = LinearLayoutManager(this@PullRequestActivity)
    var adapterRepository = PullResquestAdapter(pullRequestArrayList, this@PullRequestActivity)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_pull_request)

        owner = intent.getStringExtra(constans.owner)
        repositoryName = intent.getStringExtra(constans.repository)

        toolbar.title = repositoryName
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setDisplayShowHomeEnabled(true)

        if (savedInstanceState != null) {
            pullRequestArrayList.addAll(savedInstanceState.getParcelableArrayList(constans.listPullRequestInstance))
            owner = savedInstanceState.getString(constans.owner)
            repositoryName = savedInstanceState.getString(constans.repository)
        }else {
            requestPull()
        }

        rv_pull_request.apply {
            layoutManager = linearLayoutManager
            adapter = adapterRepository
            updateList()
        }
    }

    private fun updateList() {
        rv_pull_request.adapter?.notifyDataSetChanged()
    }

    private fun requestPull (){
        setProgressBar(true)
        val callRespository = retrofit.pullRequestService.listPullRequests(owner, repositoryName)
        callRespository.enqueue(object : Callback<List<PullRequest>> {
            override fun onResponse(call: Call<List<PullRequest>>, response: Response<List<PullRequest>>) {
                val listPullRequest = response.body()!!
                setProgressBar(false)
                pullRequestArrayList.addAll(listPullRequest)
                updateList()
            }

            override fun onFailure(call: Call<List<PullRequest>>, t: Throwable) {
                Log.e(constans.msgError, t.message)
                setProgressBar(false)
            }
        })
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelableArrayList(constans.listPullRequestInstance, pullRequestArrayList)
        outState.putString(constans.owner, owner)
        outState.putString(constans.repository, repositoryName)
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return super.onSupportNavigateUp()
    }

    override fun onClick(position: Int) {
        val pullRequest = pullRequestArrayList[position]
        val url = pullRequest.html_url
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
        startActivity(intent)

    }

    fun setProgressBar(status: Boolean){
        if(status) {
            pb_loading_pull.visibility = View.VISIBLE
        }else {
            pb_loading_pull.visibility = View.GONE
        }
    }

}
